### PR TITLE
Fix: effectiveWeight should be calculated out of raw capacity, not weighted capacity

### PR DESCRIPTION
**Pull request by danoctavian**
## Context

Issue: https://github.com/NexusMutual/smart-contracts/issues/829

The bug in StakingProducts._getEffectiveWeight affects the computation:

```    uint actualWeight = Math.min(totalAllocation * WEIGHT_DENOMINATOR / totalCapacity, type(uint16).max);```

The `totalCapacity` parameter coming out of `stakingPool.getTrancheCapacities` is multiplied and scaled down in accordance to the targetWeight. However, this is not the way to compute the `actualWeight` which should be calculated with respect to the raw capacity, without involving that product's target weight.

## Changes proposed in this pull request

The proposed change is to add a `productTargetWeight` parameter to StakingPool.getTrancheCapacities in order to allow obtaining raw capacity unaffected by productTargetWeight by setting it to 100 in this case. All other usage cases just read the actual value for the the product target weight and pass it in.

## Test plan

Wrote a set of integration tests specifically for `StakingProducts.recalculateEffectiveWeightsForAllProducts`.

Testing of the whole upgrade and of the fixing of current production broken values is done on the fork test.

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/modules/staking/StakingPool.sol [hunk 1]
```diff
@@ -849,11 +849,11 @@ contract StakingPool is IStakingPool, Multicall {
   }
 
   function getTrancheCapacities(
-    uint productId,
     uint firstTrancheId,
     uint trancheCount,
     uint capacityRatio,
-    uint reductionRatio
+    uint reductionRatio,
+    uint productTargetWeight
```
Comment (danoctavian) [inline contracts/modules/staking/StakingPool.sol:856]: reason for doing this:

Couldn't find a way to reliably reverse the calculation in StakingProducts in the case that productTargetWeight is set to 0 (information loss).

We could still expose the same function (exact same interface) without this parameter and have an internal _ getTrancheCapacities that has the parameter and is used by all other ones.

I didn't see getTrancheCapacities used anywhere yet though so i modified it directly.

File: test/integration/StakingProducts/recalculateEffectiveWeightsForAllProducts.js [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+          targetWeight: targetWeightAfterCoverBuy,
+          productId,
+        },
+      ]);
+      firstProductExpectedActiveWeight = expectedActiveWeight;
+    }
+
+    let secondProductExpectedActiveWeight;
+    {
+      // Cover inputs for cover purchase used to increase activeWeight
```
Comment (danoctavian) [inline test/integration/StakingProducts/recalculateEffectiveWeightsForAllProducts.js:236]: could refactor this block into a function, i duplicated it for now

**Review by MilGard91 — COMMENTED** [review]
Apart for the ids, lgtm

File: test/fork/recalculate-effective-weights-bug-fix.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    this.config = await getConfig.call(this);
+  });
+
+  it('should recalculate effective weight for all products in all pools', async function () {
+    const { stakingPoolFactory, stakingProducts } = this;
+
+    const poolCount = await stakingPoolFactory.stakingPoolCount();
+
+    for (let i = 0; i <= poolCount; i++) {
+      await stakingProducts.recalculateEffectiveWeightsForAllProducts(i);
```
[the remark this answers is not in the document]
Comment (danoctavian) [reply test/fork/recalculate-effective-weights-bug-fix.js:91]: thanks good point

**Review by MilGard91 — APPROVED** [review]

File: test/fork/recalculate-effective-weights-bug-fix.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
+const { ethers } = require('hardhat');
```
Comment (danoctavian) [inline test/fork/recalculate-effective-weights-bug-fix.js:1]: i made a separate file from @kyledewy  to just test this particular upgrade but reused his verification logic.

Happy to adjust if you think there should just be 1.

**Review by roxdanila — CHANGES_REQUESTED** [review]
I think we can find a cleaner solution, this is a bit messy imo

File: contracts/modules/staking/StakingProducts.sol [hunk 5]
```diff
@@ -328,15 +336,10 @@ contract StakingProducts is IStakingProducts, MasterAwareV2, Multicall {
...
-    uint[] memory trancheCapacities = stakingPool.getTrancheCapacities(
-      productId,
-      block.timestamp / TRANCHE_DURATION, // first active tranche id
-      MAX_ACTIVE_TRANCHES,
-      globalCapacityRatio,
-      capacityReductionRatio
-    );
-
-    uint totalCapacity = Math.sum(trancheCapacities);
+    uint activeStake = stakingPool.getActiveStake();
```
Comment (danoctavian) [inline contracts/modules/staking/StakingProducts.sol:339]: fix implemented as @shark0der  proposed.

**Review by shark0der — APPROVED** [review]

**Review by roxdanila — APPROVED** [review]
🚢

**Pull request merged**